### ISSUE TITLE
Support Multi-Tenancy

### ISSUE THREAD

**Issue by prabirshrestha** [dgraph-io/dgraph#2693]
### What you wanted to do
Create multiple db/schema on the same server such as tradition dbs (Postgres/MySql/SqlServer). Useful for personal VMs and raspberry pi where I can replace traditional db with dgraph and run in only on server.

### What you actually did
Add prefix.

### Why that wasn't great, with examples
cumbersome because now I need to create variables for prefix. If I accidental drop schema, I drop everything.

### Any external references to support your case
`CREATE DATABASE foodatabase` or `CREATE SCHEMA fooschema`.
Would love to have something like `localhost:8080/alter/foodatabase`. If `foodatabase` is not provided it would default to existing behavior.

**Comment by manishrjain** [comment]
The advantage of a graph DB is that multiple data sources can be combined together into one, and queried across. Given that benefit, having the division of a database is at best a low priority feature request.

**Comment by brianbroderick** [comment]
There are many reasons to have multiple databases; for example, it's typical to have a dev, test, and prod environment with their respective databases. This makes it so the test database can be recreated before each test run. Right now, the only way to accomplish this is to either have multiple instances of Dgraph running, or to add a prefix to all predicates.  

Adding a prefix complicates queries like this:  `&api.Operation{DropAll: true}`, which I run before any tests. It would also complicate Go structs when determining the right predicate values in JSON.

It's also typical to work on many micro services at a time, but these micro services should not have any chance of data colliding with each other; they should be completely isolated. It doesn't seem realistic to have 10+ instances of DGraph running at the same time on a laptop (5+ micro services, each with a dev and test environment)

Lastly, having multiple database support will help people transitioning from an RDMS world to have an easier time making the switch.

**Comment by liqweed** [comment]
A major reason for us multiple databases is such an important factor is multitenancy. We intend to implement multitenancy with a database schema per account. That makes data isolation a lot easier (which includes removing an account for example) without provisioning and maintaining thousands of database servers. Implementing multitenancy in dgraph as a schema per account is even a stronger case in my mind since it lacks any mid-level namespace to segment the data (like tables in SQL/Cassandra or collections in MongoDB). That leaves very few options to go about segmenting the data effectively. 

I agree with @brianbroderick - we implemented a microservices approach and one of the services is currently using dgraph. We avoid using dgraph for any other service since it would involve automation complexity which we find hard justifying. Had it been any easier to work with more than a single schema, dgraph usage would certainly proliferate in our case.

**Comment by romshark** [comment]
Support for multiple isolated databases on a single server would significantly increase [API test](https://github.com/romshark/dgraph_graphql_go/tree/v1.1.1/apitest) execution speed which is [currently over 226 seconds](https://travis-ci.org/romshark/dgraph_graphql_go/jobs/536398668#L812) since all tests need to be executed serially! If I had the guarantee of isolated databases I could setup a database instance for each test individually allowing API tests to run in parallel. It'd theoretically be possible to go from 226s to under 10s (which is huge!)

I could do it myself with graph namespacing, but that'd be very error prone since there's no isolation guarantees, one test could start mutating another tests's database leading to a big mess.

I hope this feature will be implemented soon!

**Comment by aoighost** [comment]
This would also be extremely useful for my use case as well using dgraph to support multiple workspaces.

**Comment by romshark** [comment]
@aoighost 

> Also, the ability to reference nodes and create relationships across databases/workspaces would be useful as well.

It would be the opposite of useful. If you have relationships across "databases" you have a single database. The multi-database feature is about isolation such that one database is physically isolated from another yet maintained by the same process for convenience.

**Comment by aoighost** [comment]
@romshark good point

**Comment by campoy** [comment]
Whoa, this is a popular request!

OK, we'll be working on this and seeing whether it can be part of our next release v1.2 expected to be released end of September.

**Comment by AgentZombie** [comment]
This might be a good place for the label field in n-quads.

**Comment by campoy** [comment]
Hi there @AgentZombie,

Could you explain what you mean by "the label field in n-quads"?

**Comment by AgentZombie** [comment]
Sorry. I was speaking specifically about the graph label field in RDF n-quads. dgraph specifically reads RDF n-quads as a superset of n-triples but doesn't use the fourth value to specify a named graph.

From https://www.w3.org/TR/n-quads/#sec-intro

> The simplest statement is a sequence of (subject, predicate, object) terms forming an RDF triple and an optional blank node label or IRI labeling what graph in a dataset the triple belongs to, all are separated by whitespace and terminated by '.' after each statement.

This was referenced [here](https://discuss.dgraph.io/t/docs-mention-n-quads-but-not-named-graphs-are-there-named-graphs/3980), #1143, and probably other places.

**Comment by campoy** [comment]
I wasn't aware of that, and it does make sense to consider it as part of our support for multi-tenancy.

Thanks, @AgentZombie

**Comment by Willem520** [comment]
I think it will be a great feature

**Comment by aoighost** [comment]
I'd appreciate it if this were not an enterprise feature. I'm trying to build a app that uses dgraph on the backend as a graph store and multi tenancy would make it a lot easier to build without having to spin up a new docker instance for each workspace. Enterprise only would kill the use of that feature for me.

**Comment by seanlaff** [comment]
We're attempting to use one big dgraph instance to serve many discrete customers and need data isolation. This would be a great feature for us.

In the meantime we've been experimenting with putting a `tenant` predicate on every entity- however I worry that this might have some performance drawbacks since every query we send into dgraph has to be a `tenant = x` query, followed by a @filter of what the end user actually wanted.

From how I understand how dgraph does query planning, I think this means all my queries can only be as fast as that original `tenant = x` lookup (which hits millions of documents), right? (Since I always need to start at the tenant predicate *and then* filter)

**Comment by hubyhuby** [comment]
+1
Managing multiple customers on one server requires clear separation. Mistakes happen all the time ;) You don t want your "sensitive customer infos" to leak into another database. You just don t want to take the chance. 
If you are doing a SAAS you want to be able to duplicate things aswell easily. Beeing able to maintain things across 1 000 customers / DBs. 
Moreover, In Europe with GDPR reglementation, it is getting more and more restrictive (for the good).

**Comment by hubyhuby** [comment]
We are evaluating / prototyping further the use of dgraph. 
For us the minimal set to evaluate Dgraph, requires 3 environments as per the regular dev pipeline : Development / Staging / Production.

Further more the GDPR constrains my company in Europe to partition the data.
A database without Multi-Tenancy feature is a No Go for most companies in Europe.
Even an academic project in Europe cannot use the community edition if they use some kind of personal data (As you cannot tell who can access the data precisely).

As a core DB feature, I believe it should be part of the community edition.

**Comment by cosmotek** [comment]
Another upvote for this feature :D

**Comment by ChStark** [comment]
Another upvote for this feature, it can also help Dgraph Labs to launch their own Dgraph As A Service easier

**Comment by dvaldivia** [comment]
This feature was marked for the 1.2 milestone but I don’t see it in the change log of the 1.2 release, did this feature not made it?

**Comment by sleto-it** [comment]
Hi @dvaldivia,

Thanks for your question. Correct, it has not been included in v1.2, but it is in our [roadmap](https://github.com/dgraph-io/dgraph/issues/4724) and we are currently working on it

**Comment by Relaxe111** [comment]
Hello, it looks like multi-tenancy support is added to Enterprise feature. I would ask to reconsider this to make it open source or to make a more flexible pricing plan. Locking this feature in Enterprise version could be a big barrier in widely adoptoption of dgraph, at least in EU it will be for sure.

**Comment by sinnergarden** [comment]
Another upvote for this feature :D

**Comment by iluminae** [comment]
To further the discussion on this issue, I would like to bring up that using the RFC for [RDF-NQUADs](https://www.w3.org/TR/n-quads/) we see the field that was added to the original RDF triple was the _graphLabel_. 

> This RDF triple is added to the graph labeled by the production graphLabel, if no graphLabel is present the triple is added to the RDF datasets default graph.

This graph label seems like a very idiomatic way to ingest data for "different graphs", thus providing a functional separation layer here.

In fact, dgraph uses the NQUAD format during ingestion and even allows setting the graph [Label](https://pkg.go.dev/github.com/dgraph-io/dgo/v2@v2.2.0/protos/api?tab=doc#NQuad) on the NQUAD used for mutations in the public gRPC API.

I can see a backwards-compatible API change where a new parameter is added to the public API to specify graphLabel at query time, falling back to the default graph.

Obviously true multi-tenancy would require more strict ACL coordination with this graphLabel, as well as the Types belonging to individual 'graphs' - but as a first step, the application layer on top of dgraph could provide that validation before a query is sent to dgraph.

Since the Q1 label on the roadmap is obviously not happening for this feature, can we get it updated to reflect the actual expected feature scheduling?

Thanks!

**Comment by minhaj-shakeel** [comment]
**Github issues have been deprecated.** 
This issue has been moved to [discuss](https://discuss.dgraph.io/t/8756). You can follow the conversation there and also subscribe to updates by changing your notification preferences.

 <img src="https://github.com/dgraph-io/benchmarks/blob/master/images/SetNotifications.png?raw=true" alt="drawing" width="40%"/>
